docs/design/README.md

```diff
@@ -534,6 +534,149 @@ After installing packages, the result of `dnf list installed` will be output to
 This will allow maintainers to audit which kit each package was taken from.
 The rest of the variant image creation steps proceed the same way that they do now.
 
+### Build Algorithm Changes
+
+Currently, the build system relies on a flat directory of RPMs.
+It relies on Cargo to build packages so that dependent RPMs are in the directory when needed.
+When creating a variant image, all RPMs are available, but the yum install command can follow package dependencies to install the right things.
+
+When creating a kit, we won't have the yum install command to help us figure out all the right packages.
+We need to change the build system so that the dependency graph can be used to put the right packages into a kit.
+
+For starters, we will no longer rely on a flat structure in `build/rpms`.
+We will put the output of each `package` build into a subdirectory using the Cargo package's name.
+
+For example, when building `glibc/Cargo.toml` we will put the output in `build/rpms/glibc`.
+Note that more than one RPM file can result from a package build.
+The subdirectory maps the resultant RPM files to the Cargo package that produced them.
+
+At each `build-package` invocation, a distinct, flat RPM directory will be created containing the packages needed for the current `build-package` invocation.
+This will be done by traversing the Cargo dependency graph and making a list of packages and kits that are needed.
+The kits will be unpacked and the direct package dependencies will be copied from their `build/rpms` subdirectories.
+This will require us to pass a list of the packages needed for the build to the dockerfile, for example:
+
+```
+PACKAGES=pkg-a pkg-b pkg-c
+```
+
+This list needs to be complete and include only "pure" package dependencies.
+That is, it should not include any package names that are "brought in" by a kit dependency.
+The packages found in each of those subdirectories, e.g. `build/rpms/pkg-a`, will be copied to a flat RPM directory for the build.
+
+Additionally, the `build-package` command needs to know the list of kits required:
+
+```
+KITS=core-kit aws-kit
+```
+
+Each of these needs to be unpacked and have its RPMs copied out to the flat RPM directory.
+For example, these would be unpacked: `build/kits/core-kit.tar`, `build/kits/aws-kit.tar`.
+
+The `build-kit` requirements are slightly simplified.
+Only the list of "pure" package dependencies is required.
+When constructing a kit, buildsys does not need to know what other kits the current kit depends on.
+
+#### Dependency Traversal
+
+The simplest way to get a list of "pure" package dependencies and a list of kits is to traverse the dependency graph twice.
+
+In the first traversal, kit dependencies are ignored and not followed, resulting in only the "pure" package dependencies.
+In the second traversal, all nodes are followed and package dependencies are filtered out from the final list, resulting in a list of all required kits.
```

Comment:
I am still confused on why the two traversals are needed, can we clarify this need and why we can't just ensure we have the kit metadata as pregenerated. (As in when the kit is built it also generates its metadata.json which would list any dependent kits. Then gathering the total set of needed kits wouldn't be a scan each package problem, potentially.

Reply:
> I am still confused on why the two traversals are needed, can we clarify this need

The algorithm design could clarify that this is "conceptually" the simplest way. Obviously you can optimize it for a single traversal, but that would require some gnarlier state tracking since you would have to switch modes when crossing a kit boundary and put those packages in a different collection.